File: src/main/java/com/typelead/intellij/plugin/eta/lang/psi/EtaFile.java
package com.typelead.intellij.plugin.eta.lang.psi;

import com.intellij.extapi.psi.PsiFileBase;
import com.intellij.openapi.fileTypes.FileType;
import com.intellij.psi.FileViewProvider;
import com.typelead.intellij.plugin.eta.lang.EtaFileType;
import com.typelead.intellij.plugin.eta.lang.EtaLanguage;
import com.typelead.intellij.plugin.eta.resources.EtaIcons;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;

public final class EtaFile extends PsiFileBase {

    public EtaFile(@NotNull FileViewProvider viewProvider) {
        super(viewProvider, EtaLanguage.INSTANCE);
    }

    @NotNull
    @Override
    public FileType getFileType() {
        return EtaFileType.INSTANCE;
    }

    @Nullable
    @Override
    public Icon getIcon(int flags) {
        return EtaIcons.FILE;
    }

    @Override
    public String toString() {
        return "Eta file";
    }
}

File: src/main/eta/FFI/Com/TypeLead/IntelliJ/Plugin/Eta/Lang/Psi/EtaFile.hs
module FFI.Com.TypeLead.IntelliJ.Plugin.Eta.Lang.Psi.EtaFile where

import P

import FFI.Com.IntelliJ.Psi.FileViewProvider
import FFI.Com.IntelliJ.Psi.PsiFile

data {-# CLASS "com.typelead.intellij.plugin.eta.lang.psi.EtaFile" #-}
  EtaFile = EtaFile (Object# EtaFile)
  deriving Class

type instance Inherits EtaFile = '[PsiFile]

foreign import java unsafe "@new" newEtaFile
  :: FileViewProvider -> Java a EtaFile
